spring-session/src/main/java/org/springframework/session/data/gemfire/GemFireOperationsSessionRepository.java

```diff
@@ -0,0 +1,352 @@
+
+	// GemFire OQL query used to look up Sessions by principal name.
+	protected static final String FIND_SESSIONS_BY_PRINCIPAL_NAME_QUERY =
+		"SELECT s FROM %1$s s WHERE s.principalName = $1";
+
+	/**
+	 * Constructs an instance of GemFireOperationsSessionRepository initialized with the required GemfireOperations
+	 * object used to perform data access operations to manage Session state.
+	 *
+	 * @param template the GemfireOperations object used to access and manage Session state in GemFire.
+	 * @see org.springframework.data.gemfire.GemfireOperations
+	 */
+	public GemFireOperationsSessionRepository(GemfireOperations template) {
+		super(template);
+	}
+
+	/**
+	 * Looks up all the available Sessions tied to the specific user identified by principal name.
+	 *
+	 * @param principalName the principal name (i.e. username) to search for all existing Spring Sessions.
+	 * @return a mapping of Session ID to Session instances.
+	 * @see org.springframework.session.ExpiringSession
+	 */
+	@Override
+	public Map<String, ExpiringSession> findByPrincipalName(String principalName) {
+		SelectResults<ExpiringSession> results = getTemplate().find(String.format(
+			FIND_SESSIONS_BY_PRINCIPAL_NAME_QUERY, getFullyQualifiedRegionName()), principalName);
+
+		Map<String, ExpiringSession> sessions = new HashMap<String, ExpiringSession>(results.size());
+
+		for (ExpiringSession session : results.asList()) {
+			sessions.put(session.getId(), session);
+		}
+
+		return sessions;
+	}
+
+	/**
+	 * Constructs a new {@link ExpiringSession} instance backed by GemFire.
+	 *
+	 * @return an instance of {@link ExpiringSession} backed by GemFire.
+	 * @see org.springframework.session.data.gemfire.GemFireOperationsSessionRepository.GemFireSession
+	 * @see org.springframework.session.ExpiringSession
+	 */
+	@Override
+	public ExpiringSession createSession() {
+		return GemFireSession.create(getMaxInactiveIntervalInSeconds());
+	}
+
+	/**
+	 * Gets a copy of an existing, non-expired {@link ExpiringSession} by ID.  If the Session is expired,
+	 * then it is deleted.
+	 *
+	 * @param sessionId a String indicating the ID of the Session to get.
+	 * @return an existing {@link ExpiringSession} by ID or null if not Session exists.
+	 * @see org.springframework.session.ExpiringSession
+	 * @see #delete(String)
+	 */
+	@Override
+	public ExpiringSession getSession(String sessionId) {
+		ExpiringSession storedSession = getTemplate().get(sessionId);
+
+		if (storedSession != null) {
+			if (storedSession.isExpired()) {
+				delete(storedSession.getId());
+			}
+			else {
+				return GemFireSession.from(storedSession);
+			}
+		}
+
+		return null;
+	}
+
+	/**
+	 * Saves the specified {@link ExpiringSession} to GemFire.
+	 *
+	 * @param session the {@link ExpiringSession} to save.
+	 * @see org.springframework.session.ExpiringSession
+	 */
+	@Override
+	public void save(ExpiringSession session) {
```

Comment:
This is something the GemFire team did with the HTTP Session Management module and something we could implement on behalf of the user.  In GemFire, it is a simple "flag" to flush on write.